components/ipfs/pin/ipfs_local_pin_service.cc

```diff
@@ -0,0 +1,284 @@
+        } else {
+          verification_pased = false;
+          list->EraseValue(base::Value(prefix_));
+        }
+        if (list->empty()) {
+          update_dict.Remove(cid);
+        }
+      }
+    }
+    std::move(callback_).Run(verification_pased);
+  } else {
+    std::move(callback_).Run(absl::nullopt);
+  }
+}
+
+GcJob::GcJob(PrefService* prefs_service,
+             IpfsService* ipfs_service,
+             GcCallback callback)
+    : prefs_service_(prefs_service),
+      ipfs_service_(ipfs_service),
+      callback_(std::move(callback)) {}
+GcJob::~GcJob() {}
+
+void GcJob::Start() {
+  ipfs_service_->GetPins(
+      absl::nullopt, kRecursiveMode, true,
+      base::BindOnce(&GcJob::OnGetPinsResult, base::Unretained(this)));
+}
+
+void GcJob::OnGetPinsResult(bool status, absl::optional<GetPinsResult> result) {
+  std::vector<std::string> cids_to_delete;
+  if (status && result) {
+    const base::Value::Dict& dict = prefs_service_->GetDict(kIPFSPinnedCids);
+    for (const auto& pair : result.value()) {
+      const base::Value::List* list = dict.FindList(pair.first);
+      if (!list || list->empty()) {
+        cids_to_delete.push_back(pair.first);
+      }
+    }
+
+    if (!cids_to_delete.empty()) {
+      ipfs_service_->RemovePin(
+          cids_to_delete,
+          base::BindOnce(&GcJob::OnPinsRemovedResult, base::Unretained(this)));
+    } else {
+      std::move(callback_).Run(true);
+    }
+  } else {
+    std::move(callback_).Run(false);
+  }
+}
+
+void GcJob::OnPinsRemovedResult(bool status,
+                                absl::optional<RemovePinResult> result) {
+  if (status && result) {
+    std::move(callback_).Run(true);
+  } else {
+    std::move(callback_).Run(false);
+  }
+}
+
+IpfsLocalPinService::IpfsLocalPinService(PrefService* prefs_service,
+                                         IpfsService* ipfs_service)
+    : prefs_service_(prefs_service), ipfs_service_(ipfs_service) {
+  ipfs_base_pin_service_ =
+      std::make_unique<IpfsBasePinService>(prefs_service_, ipfs_service_);
+  base::SequencedTaskRunnerHandle::Get()->PostDelayedTask(
+      FROM_HERE,
+      base::BindOnce(&IpfsLocalPinService::AddGcTask, base::Unretained(this)),
+      base::Minutes(1));
+}
+
+IpfsLocalPinService::IpfsLocalPinService() {}
+
+void IpfsLocalPinService::SetIpfsBasePinServiceForTesting(
+    std::unique_ptr<IpfsBasePinService> service) {
+  ipfs_base_pin_service_ = std::move(service);
+}
+
+IpfsLocalPinService::~IpfsLocalPinService() {}
+
+void IpfsLocalPinService::AddPins(const std::string& prefix,
+                                  const std::vector<std::string>& cids,
+                                  AddPinCallback callback) {
+  ipfs_base_pin_service_->AddJob(std::make_unique<AddLocalPinJob>(
+      prefs_service_, ipfs_service_, prefix, cids,
+      base::BindOnce(&IpfsLocalPinService::OnAddJobFinished,
+                     base::Unretained(this), std::move(callback))));
+}
```

Comment:
It is quite hard to track how lifetimes of callback and this correlate. Better pass it by weak ptr and be safe.